src/OauthTokenFileStorage.php

```diff
@@ -19,16 +19,21 @@
 
 namespace Drupal\apigee_edge;
 
-use Apigee\Edge\HttpClient\Plugin\Authentication\OauthTokenStorageInterface;
-use Drupal\apigee_edge\Form\AuthenticationForm;
+use Drupal\apigee_edge\Exception\OauthTokenStorageException;
+use Drupal\Core\Config\ConfigFactoryInterface;
+use Drupal\Core\Logger\LoggerChannelInterface;
+use Drupal\Core\Site\Settings;
 
 /**
- * Storing and returning OAuth access token data.
- *
- * @internal The implementation may change before the first stable release.
+ * Stores OAuth token data in a file.
  */
 final class OauthTokenFileStorage implements OauthTokenStorageInterface {
```

Comment:
Why is this service marked as final?

Reply:
Our service _implementations_ are not for extension. If someone would like to define its own implementation for a service s/he should create a new class and implement the service's interface definition. That describe the requirements of a service and not a concrete class implementation.
By marking a class as final we are not making a class non-reusable, we just driving the developers to the right directions, forcing them to prefer encapsulation over inheritance when they would like build something in the top of that class.
Thanks for keeping BC between releases will be much easier for us. We can change the internal behavior of our classes (property names, methods, etc.) however we want until we keep or class implementations compatible with the interfaces that they implements.

Reply:
When you define a service you should keep in my that it either gets **decorated** or **overridden** by external developers if it is needed. If you define an interface for a service you can force developers to keep the contract the the service's original implementation describes. Also, do not forget you should use interfaces in type hints wherever you require a service.